src/qibo/validation/gate_set_tomography.py

```diff
@@ -0,0 +1,2089 @@
+                GST_nogates_qibo_2qb = "GST_nogates_qibo_2qb"
+                if not os.path.exists(GST_nogates_qibo_2qb):
+                    os.makedirs(GST_nogates_qibo_2qb)
+
+            elif class_qubit_gate is not None:
+                filedirectory = "GST_gates_qibo_2qb/"
+                fname = "%s gate (k=%.2d,j=%.2d) %s_basis.txt" % (
+                    class_qubit_gate,
+                    k,
+                    j,
+                    idx_basis[j],
+                )
+                GST_gates_qibo_2qb = "GST_gates_qibo_2qb"
+                if not os.path.exists(GST_gates_qibo_2qb):
+                    os.makedirs(GST_gates_qibo_2qb)
+
+            # Save counts into a matrix. 0th column: word; 1st column: count value
+            # --------------------------------------------------------------------
+            fout = filedirectory + fname
+            fo = open(fout, "w")
+            for v1, v2 in counts.items():
+                fo.write(str(v1) + "  " + str(v2) + "\n")
+            fo.close()
+
+        # Find value for matrix_jk
+        # -------------------------
+
+        expectation_signs = np.matrix(
+            [[1, 1, 1, 1], [1, -1, 1, -1], [1, 1, -1, -1], [1, -1, -1, 1]]
+        )
+
+        matrix = np.zeros((2**2, 2))
+        for ii in range(0, 2**2):
+            word = bin(ii)[2:]
+            matrix[ii, 0] = word
+
+        for v1, v2 in counts.items():
+            # print(v1, v2)
+            row = int(v1, 2)
+            col = v2
+            matrix[row, 1] = col
+
+        sorted_matrix = sort_counts_2qb(matrix)
+        probs = sorted_matrix[:, 1] / np.sum(sorted_matrix[:, 1])
+        sorted_matrix = np.column_stack((sorted_matrix, probs))
+
+        k = k + 1
+        j = j + 1
+
+        if j == 1:
+            sorted_matrix = np.column_stack((sorted_matrix, expectation_signs[:, 0]))
+        elif j == 2:
+            sorted_matrix = np.column_stack((sorted_matrix, expectation_signs[:, 1]))
+        elif j == 3:
+            sorted_matrix = np.column_stack((sorted_matrix, expectation_signs[:, 1]))
+        elif j == 4:
+            sorted_matrix = np.column_stack((sorted_matrix, expectation_signs[:, 1]))
+        elif j == 5:
+            sorted_matrix = np.column_stack((sorted_matrix, expectation_signs[:, 2]))
+        elif j == 9:
+            sorted_matrix = np.column_stack((sorted_matrix, expectation_signs[:, 2]))
+        elif j == 13:
+            sorted_matrix = np.column_stack((sorted_matrix, expectation_signs[:, 2]))
+        else:
+            sorted_matrix = np.column_stack((sorted_matrix, expectation_signs[:, 3]))
+
+        temp = np.array(sorted_matrix[:, 2]) * np.array(sorted_matrix[:, 3])
+        sorted_matrix = np.column_stack((sorted_matrix, temp))
+
+        expectation_val = np.sum(temp)
+
+        return expectation_val
+
+    nqubits = 2
+    matrix_jk_2qb = np.zeros((4**nqubits, 4**nqubits))
+    for k in range(0, 16):
+        if k == 0:  # |0> |0>
```

Comment:
I would probably define a helper function that given k returns the needed circuit, it would make it easier to follow the code.

Reply:
Thanks Andrea, I will look into building a helper function for k.

Reply:
This is done.